### ISSUE TITLE
Is it possible to upload external files into ./files/ directory using TiddlyWiki?

### ISSUE THREAD

**Issue by pimgeek** [Jermolene/TiddlyWiki5#4401]
In this [link](https://tiddlywiki.com/#Using%20the%20integrated%20static%20file%20server) I learned that I can link to `./files/xxx.yyy` files in tiddlers using `[ext[./files/xxx.yyy]` syntax on Node.js backed TiddlyWiki website. 😃

But if I use TiddlyWiki's drag-and-drop / import files action to upload external files (not only image files, but also video files and pdf files), it always creates big tiddler with embedded binary data. 😑

So my question is, is there a way to **upload external files directly into `./files/` directory (and auto-create a skinny tiddler pointing to that file)** using TiddlyWiki? I've searched throughout GitHub Issues / Google Groups / Official Documents / Dynalist Tool Map outlines without finding a solution... 😥

**Comment by Jermolene** [comment]
Hi @pimgeek there is no support at present for uploading files from the browser directly to the `./files/` area. It is something that I'd like to add at some point.

**Comment by pimgeek** [comment]
@Jermolene Thanks for your reply, currently I consider writing a file system watching script to automatically create new tiddlers pointing to ./files/xxx , but afraid it will confuse the tiddler updating mechanism. 😅

**Comment by joshuafontany** [comment]
Bob and Bib.exe both have this option (along with serving multiple wikis,
multiple users, & other great features).

https://github.com/OokTech/TW5-Bob/blob/master/Documentation/ServerMedia.tid

**Comment by pimgeek** [comment]
> Bob and Bib.exe both have this option (along with serving multiple wikis, multiple users, & other great features). https://github.com/OokTech/TW5-Bob/blob/master/Documentation/ServerMedia.tid

Thanks for your recommendation, I will have a try. 👌

**Comment by Arlen22** [comment]
@pimgeek This is a very simple feature to implement and I thought it was already implemented in TiddlyWiki. 

Actually, upon further investigation, it apparently already is. This should already work.

https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/server/routes/get-file.js

Create a `files` folder in the wiki data folder (i.e. beside the `tiddlers` folder), then you should be able to access any files there directly. Start with a simple file with no spaces (like `test.txt`) to make sure it works and then go from there. If you find any bugs feel free to open another issue.

**Comment by Arlen22** [comment]
You would access the file at `http://localhost:8080/files/text.txt` or if using TiddlyServer `http://localhost:8080/path/to/wiki/files/test.txt`.

**Comment by Jermolene** [comment]
HI @Arlen22 I think @pimgeek wants to be able to upload files into the `/files` folder directly from the browser, which is still not currently supported.

**Comment by linonetwo** [comment]
But why do we need this? Just import them to the /tiddlers folder, and using lazy-image feature is enough. (Which is enabled by default in TidGi desktop)

**Comment by pimgeek** [comment]
> using lazy-image feature is enough

From what you said, I guess there are some newly updated features on how TiddlyWiki process imported image / video / pdf tiddlers, but I didn't notice and try.

I did notice LazyLoading mechanism on TiddlyWiki official documents, but that's several years ago, and at that time `lazy-all` would break TiddlyMap and `lazy-images` didn't seem to decrease page loading time dramatically. 

Maybe I was wrong about `lazy-images` command line option's real effect, I will try using it again by importing very large files to see if it does work as expected. 🤓

**Comment by arunnbabu81** [comment]
@pimgeek I use [tiddlyclip plug-in](https://github.com/buggyj/tiddlyclip-plugin), [tiddlyclip](https://github.com/buggyj/tiddlyclip/releases/tag/0.1.8) and [save media](https://github.com/buggyj/savemedia) add-ons by @buggyj to clip images (using screenshots in firefox or snap feature of tiddlyclip add on) directly from websites or pdf into tiddlers in this format `[img[.relativepath]]` on clipping. Images get saved automatically into a `media` folder in the subdirectory of the singe file wiki. Multiple images can be clipped into a single tiddler in this way. I have been using this for the past many months. Till TW support this natively, this is the best option I have found.

**Comment by pmario** [comment]
>But why do we need this? Just import them to the /tiddlers folder, and using lazy-image feature is enough. (Which is enabled by default in TidGi desktop)

The biggest problem with unoptimized images directly imported via TW is **size** ... I did create a little PNG screenshot and directly imported it to TW. 

Images optimized with https://squoosh.app/editor in the browser. 

|Image | File Size | Size Saved | Settings | Description
|--|--|--|--|--|
image.png | 121 KB | -- | browser default | unoptimized browser saves it
image.oxi.png | 70 KB | 42% | default | using OxiPNG optimisation
image.avif | 26 KB | 79% | quality 40% | using AVIF format

<details><summary>Click to expand the screenshot, that contains mainly whitespace</summary>

![image](https://github.com/Jermolene/TiddlyWiki5/assets/374655/ea65e623-c097-4903-9f8c-38508b859c7a)

</details>

**Comment by pmario** [comment]
All Images to compare them

<details><summary>121 KB</summary>

![image](https://github.com/Jermolene/TiddlyWiki5/assets/374655/457ad29e-c9f8-457e-8990-90f68bd7eb8c)

</details>


<details><summary>70 KB</summary>

![image oxi](https://github.com/Jermolene/TiddlyWiki5/assets/374655/e0c97b36-1eea-4650-afa6-707369ce6fdc)

</details>

Sorry AVIF cannot be uploaded atm.

[image.avif.zip](https://github.com/Jermolene/TiddlyWiki5/files/11732334/image.avif.zip)

**Comment by linonetwo** [comment]
@pimgeek You can also try https://saqimtiaz.github.io/tw5-file-uploads/ to see if this issue can be closed.

**Comment by pmario** [comment]
>And you can already do it now with programs like MacOS Automator.

On windows and Linux systems?

**Comment by linonetwo** [comment]
programs like MacOS Automator is widely used, there are many alternatives

**Comment by pimgeek** [comment]
@linonetwo I've finished evaluting the above solutions:

1. `lazy-image` solution: it works as you said, big image files are not loaded when I open the website, they are only loaded when I open the image file tiddler. (but for mp4 file tiddler, I need to use `lazy-all` option)
2. `File Uploads plugin`: it doesn't provide a filesystem uploader sub-plugin, that means I cannot use it to upload to `./files/` or `./media/` dir, so it is not currently a solution
3. `Automator` or `AutoHotKey`: it is a practical solution. I myself had written background python script to do something similar --- whenever a file is copied to `./files/` or `./media/` dir, the script would automatically create a skinny tiddler with `_canonical_uri`, but it feels clumsy when put into production environment.
4. `tiddlyclip plug-in, tiddlyclip and save media`: it's not as simple as `lazy-image` solution, and it does not support mp4 / pdf files upload either.

In summary, the most promising solution seems to be `lazy-all` option (recently I don't use TiddlyMap anymore, so no error message is popped up). I will use it in production environment for some time. Since this solution is essentially the same as uploading the file to Node.js server without increasing loading-time, so I will close this issue.

**Comment by pimgeek** [comment]
> I only use lazy-all in my wiki on TidGi locally, not on a public site

I will use it in my home network, with another user. This is not a public website, but I still view it as production environment, because the other user does not have technical background, and will help discover some inconveniences which were previously omitted by me.

**Comment by pmario** [comment]
>I will use it in my home network, with another user. 

Which OSes do you use? 

If you work with the local file systems, you can handle the access rights on the OS level, so your server setup can be much more bare bones. 

It may be an option to use a local WebDav server, which can deal with single file wikis and it works with Saq's file-upload plugin. There is an into-wiki at: https://saqimtiaz.github.io/sq-tw/webdav/ which describes the "wiki-farm" project a little bit. 

Saq's plugins are there: https://saqimtiaz.github.io/tw5-plugins-sandbox/

**Comment by pimgeek** [comment]
> It may be an option to use a local WebDav server, which can deal with single file wikis and it works with Saq's file-upload plugin.

Thanks for your suggestion, I managed to setup WebDAV server and later noticed that it only supports Single-File TiddlyWiki, so I searched in Discourse Forum and found this solution: https://talk.tiddlywiki.org/t/use-external-images-in-node-js/3332/4 which is also implemented by Saq. 

This **Node.js files PUT Uploader** solution is better than `lazy-images` (On some mobile browsers, images don't show up the first time I open a tiddler, I have to open it twice to make the images load). Saq's solution is still in Experimental mode, but it is closest to what I have been searching for. 😃
